### PR TITLE
Workaround for cases where MPI_Finalize is called before MPI_Comm_free

### LINKED ISSUE #2951: MPI_Comm_free() called after MPI_FINALIZE
It had been a while since I updated deal.ii, so I can't pin down when this happened, but recently I've been getting MPI errors at the end of parallel model runs:
```
*** The MPI_Comm_free() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
```
I first noticed it in Aspect, but it apparently happens in step-32 also, so I don't think it's a regression downstream. I haven't changed my Trilinos install, so it's either in deal.ii or p4est. It happens on Ubuntu, with GCC 5.4 and OpenMPI 1.10.2, and Debian testing, which I believe is on GCC 6.x (I don't have the MPI version at hand for that machine).

### PR DESCRIPTION
**Pull request by jperryhouts**
This fixes #2951 in a more elegant way than either #3018 or #3019

### CODE REVIEW

File: source/base/mpi.cc [hunk 1]
```diff
@@ -485,6 +495,12 @@ namespace Utilities
         }
 #endif
 
+#ifdef DEAL_II_WITH_P4EST
+      // now end p4est and libsc
+      // Note: p4est has no finalzie function
```
Comment (bangerth) [inline source/base/mpi.cc]: typo

**Comment by bangerth** [comment]
This looks ok to me, with the exception of the one typo.

**Comment by jperryhouts** [comment]
Fixed the typo.

**Comment by bangerth** [comment]
Thinking about it:
* Can you add that `MPI_InitFinalize` also does the sc/p4est initialization in that class's documentation
* Can you double check that all tutorial programs that use MPI really use this class?

**Comment by jperryhouts** [comment]
Comment added.

And yes, I believe all the tutorials do use MPI_InitFinalize.
```
jmp@Higgs:dealii$ for f in `grep -Rl "MPI::" examples | grep -o "step-.." | sort | uniq`; do echo -n "$f: "; grep -R MPI_InitFinalize "examples/$f" | wc -l; done
step-17: 2
step-18: 1
step-31: 2
step-32: 2
step-33: 1
step-36: 1
step-40: 2
step-41: 1
step-42: 1
step-43: 1
step-45: 1
step-48: 2
step-50: 1
step-55: 1
```

**Comment by bangerth** [comment]
Great, thanks for double checking.

I have to pack in order to leave tomorrow morning at 4:30am. I'd appreciate if someone else could look this over one more time and do the merge.

Thanks for sorting this out, @jperryhouts !

**Comment by tjhei** [comment]
Excellent, thanks! :+1:

**Pull request merged**